### PR TITLE
kbingham/rpi 6.10.y VC04_services fixes

### PR DESCRIPTION
**Pull request by kbingham**
Version 2 *COMPILE TESTED ONLY*

Now ready for closer examination and review (and testing?)

### CODE REVIEW

**Comment by kbingham** [comment]
Trying to build/test this against the bcm2711_defconfig I get the following errors
```
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/common: No such file or directory [-Werror=missing-include-dirs]
/home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/src/pcie.c:23:10: fatal error: hailo_pcie_version.h: No such file or directory
   23 | #include "hailo_pcie_version.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
```

I don't think those are related to my branch here (Presumably caused in c5f657483e) so I'll disable HAILO and continue.

File: drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c [hunk 1]
```diff
@@ -1760,63 +1757,61 @@ static int bcm2835_isp_probe_instance(struct platform_device *pdev,
...
 
-module_platform_driver(bcm2835_isp_pdrv);
+module_vchiq_driver(bcm2835_isp_drv);
 
 MODULE_DESCRIPTION("BCM2835 ISP driver");
 MODULE_AUTHOR("Naushir Patuck <naush@raspberrypi.com>");
 MODULE_LICENSE("GPL");
 MODULE_VERSION("1.0");
-MODULE_ALIAS("platform:bcm2835-isp");
+MODULE_ALIAS("vchiq:bcm2835-isp");
```
Comment (kbingham) [inline drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:1817]: Argh, this is inconsistent compared to bcm2835-codec. And also I think the upstream(/staging) drivers remove this in favour of a compatible ID table that should also be adapted in here.

**Comment by popcornmix** [comment]
On a Pi4 I get a panic on boot that is solved if I remove this line:
https://github.com/raspberrypi/linux/blob/b4937bfc429c83c1323f51724297cf109caa323e/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c#L1737

@6by9 do you know why you added that?
It appeared in [this commit](https://github.com/raspberrypi/linux/commit/b4937bfc429c83c1323f51724297cf109caa323e) and appears incongruous.

**Comment by 6by9** [comment]
> On a Pi4 I get a panic on boot that is solved if I remove this line:
> 
> https://github.com/raspberrypi/linux/blob/b4937bfc429c83c1323f51724297cf109caa323e/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c#L1737
> 
> @6by9 do you know why you added that? It appeared in [this commit](https://github.com/raspberrypi/linux/commit/b4937bfc429c83c1323f51724297cf109caa323e) and appears incongruous.

Looks like a duff conflict resolution to me - it's not in the 6.6 version of that patch (https://github.com/raspberrypi/linux/commit/fa9b3bdb56e69dc21d449d041518f8ba58563823)
Ditch it.

**Comment by popcornmix** [comment]
Okay - pushed its removal.
While we boot, with this PR, there are no /dev/video nodes being created for bcm2835-codec.
On 6.6 kernel:
```
pi@pi4:~ $ dmesg | grep bcm2835.codec
[   16.177889] bcm2835_codec: module is from the staging directory, the quality is unknown, you have been warned.
[   16.263185] bcm2835-codec bcm2835-codec: Device registered as /dev/video10
[   16.274552] bcm2835-codec bcm2835-codec: Loaded V4L2 decode
[   16.291916] bcm2835-codec bcm2835-codec: Device registered as /dev/video11
[   16.305535] bcm2835-codec bcm2835-codec: Loaded V4L2 encode
[   16.338784] bcm2835-codec bcm2835-codec: Device registered as /dev/video12
[   16.357214] bcm2835-codec bcm2835-codec: Loaded V4L2 isp
[   16.374435] bcm2835-codec bcm2835-codec: Device registered as /dev/video18
[   16.389027] bcm2835-codec bcm2835-codec: Loaded V4L2 image_fx
[   16.409554] bcm2835-codec bcm2835-codec: Device registered as /dev/video31
[   16.416599] bcm2835-codec bcm2835-codec: Loaded V4L2 encode_image
```
but no matches on 6.10.

**Comment by kbingham** [comment]
> On a Pi4 I get a panic on boot that is solved if I remove this line:
> 
> https://github.com/raspberrypi/linux/blob/b4937bfc429c83c1323f51724297cf109caa323e/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c#L1737
> 
> @6by9 do you know why you added that? It appeared in [this commit](https://github.com/raspberrypi/linux/commit/b4937bfc429c83c1323f51724297cf109caa323e) and appears incongruous.

Aha - I've just caught up to this point. But my workflow is really slow as I have to reload kernels on an SD card (now deployed gpio-jumper hack to revert to working kernel for ssh updates :D) 
... I should really get tftpboot set up for this. But now I have a pi4 with fresh install.

**Comment by kbingham** [comment]
> While we boot, with this PR, there are no /dev/video nodes being created for bcm2835-codec.

I would bet this is module autoloading issues. I'll look at this now.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 2]
```diff
@@ -4000,10 +3995,10 @@ static struct platform_driver bcm2835_v4l2_codec_driver = {
...
 
-module_platform_driver(bcm2835_v4l2_codec_driver);
+module_vchiq_driver(bcm2835_v4l2_codec_driver);
 
 MODULE_DESCRIPTION("BCM2835 codec V4L2 driver");
 MODULE_AUTHOR("Dave Stevenson, <dave.stevenson@raspberrypi.com>");
 MODULE_LICENSE("GPL");
 MODULE_VERSION("0.0.1");
-MODULE_ALIAS("platform:bcm2835-codec");
+MODULE_ALIAS("bcm2835-codec");
```
Comment (kbingham) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: yes, this should be:
`MODULE_ALIAS("vchiq:bcm2835-codec");`
Testing ...
```
kbingham@mars:~ $ lsv4l2 
v4l-subdev0: unicam
v4l-subdev1: imx219 10-0010
video0: unicam-image
video1: unicam-embedded
video10: bcm2835-codec-decode
video11: bcm2835-codec-encode
video12: bcm2835-codec-isp
video13: bcm2835-isp-output0
video14: bcm2835-isp-capture0
video15: bcm2835-isp-capture1
video16: bcm2835-isp-stats2
video18: bcm2835-codec-image_fx
video19: rpivid
video20: bcm2835-isp-output0
video21: bcm2835-isp-capture0
video22: bcm2835-isp-capture1
video23: bcm2835-isp-stats2
video31: bcm2835-codec-encode_image
```

I'll rebase, fixup and repush

**Comment by kbingham** [comment]
Camera is failing - but I don't (yet) think that's related to vchiq. Could be more about the changes that have occured to imx219 upstream or otherwise.

```
kbingham@mars:~ $ cam -c1 -C100
[0:18:48.705770346] [1009]  INFO Camera camera_manager.cpp:284 libcamera v0.2.0+46-075b54d5
[0:18:48.753482075] [1012]  WARN RPiSdn sdn.cpp:39 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[0:18:48.755214254] [1012]  WARN RPI vc4.cpp:393 Mismatch between Unicam and CamHelper for embedded data usage!
[0:18:48.755799101] [1012]  INFO RPI vc4.cpp:447 Registered camera /base/soc/i2c0mux/i2c@1/imx219@10 to Unicam device /dev/media0 and ISP device /dev/media2
Using camera /base/soc/i2c0mux/i2c@1/imx219@10 as cam0
[0:18:48.756489523] [1009]  INFO Camera camera.cpp:1183 configuring streams: (0) 800x600-XRGB8888
[0:18:48.756856585] [1012]  INFO RPI vc4.cpp:611 Sensor: /base/soc/i2c0mux/i2c@1/imx219@10 - Selected sensor format: 1640x1232-SBGGR10_1X10 - Selected unicam format: 1640x1232-pBAA
[0:18:48.775344991] [1012] ERROR V4L2 v4l2_videodevice.cpp:1906 /dev/video0[16:cap]: Failed to start streaming: Broken pipe
Failed to start capture
Failed to start camera session

```

**Comment by popcornmix** [comment]
Updated version gets codecs probed and working (e.g. h264 from kodi is hw accelerated).

If the imx219 issues is unrelated that can be solved later, so I think this PR can be merged. Thanks.

**Comment by kbingham** [comment]
@popcornmix I'm trying to build vchiq_test from https://github.com/raspberrypi/userland/ and failing.

I see 64 bit isn't supported though - Is vchiq_test packaged anywhere? Do you need any further testing from me here?

**Comment by popcornmix** [comment]
The userland repo is deprecated, so it's probably not worth trying to build things from it.
As long as vchiq is working from the kernel (and playing h264 via mmal suggests it is) then we are probably good.

**Comment by lategoodbye** [comment]
@kbingham Good catch, regarding [7cfe258](https://github.com/raspberrypi/linux/pull/6205/commits/7cfe25874f779915d0dceb745e8336b39d21fa36)

Do you plan to submit it mainline?

**Comment by kbingham** [comment]
> @kbingham Good catch, regarding [7cfe258](https://github.com/raspberrypi/linux/pull/6205/commits/7cfe25874f779915d0dceb745e8336b39d21fa36)
> 
> Do you plan to submit it mainline?

Oh yes this definitely needs posting. 
I'll try to handle it tonight so I don't lose it.

**Pull request merged**